### PR TITLE
Add group manipulation methods to group.py

### PR DESCRIPTION
**Pull request by tsalemink**
In addition to the existing `group_add_group_elements` method, this commit adds an inverse `group_remove_group_elements` method. It also Includes _add-not_ and _remove-not_ methods which use the elements _not_ in the second group for the group operation.

@rchristie could you please check my logic for the two _not_ methods, these occasionally give unexpected results.

### CODE REVIEW

File: src/cmlibs/utils/zinc/group.py
Comment (rchristie) [inline src/cmlibs/utils/zinc/group.py:35]: Same issues as for elements.

**Review by rchristie — CHANGES_REQUESTED** [review]
Some comments on the add/remove node elements/nodes.
See also separate zinc issue you've discovered.

File: src/cmlibs/utils/zinc/group.py
Comment (tsalemink) [reply src/cmlibs/utils/zinc/group.py:87]: Sorry I'm a little confused about this comment. Do you mean that if we are doing a removal and `mesh_group` is empty then we should skip performing the operations for that iteration? Why should I use `getMeshGroup` instead of `getOrCreateMeshGroup`?

Comment (rchristie) [reply src/cmlibs/utils/zinc/group.py:87]: Because the MeshGroups and NodesetGroups are created on demand, we don't want to make empty ones for no good reason in these cases:

1.  When removing from a group; if there is no current mesh/nodeset group we know we can't remove anything from it and we don't want to make an empty such group, hence use 'get' and only remove if isValid()
2. When using objects contained in a source group we should never modify it, so always use 'get' and check it isValid() before using. If it isn't valid, there is nothing to act on.

**Comment by tsalemink** [comment]
@rchristie, I have just made the changes you requested.

One more thing. You suggested removing the `size > 0` condition for `other_group` in case it's actually empty and you wanted to remove the complement of `other_group`? But, if `other_group` is empty then `other_group.isValid()` also returns False (at least in some cases). So it seems that we would also have to remove the validity check. However, based on _case 2_ you mentioned in your comment about creating empty groups, I'm not sure if this is the best idea? I have implemented this anyway and it does achieve the desired result with the empty group's complement. Let me know what you think of my current implementation.

I'll squash all the commits into one once I've got it how you want.

Thanks.

File: src/cmlibs/utils/zinc/group.py
Comment (hsorby) [inline src/cmlibs/utils/zinc/group.py:101]: I would change these lines to be:
```
conditional = other_fieldmodule.createFieldNot(other_group) if complement else other_group
```

Comment (hsorby) [reply src/cmlibs/utils/zinc/group.py:101]: (Same in mesh part)

**Review by hsorby — CHANGES_REQUESTED** [review]
I think I would also call it `conditional_group` over just `conditional`.

**Comment by hsorby** [comment]
Nice! Looks good to me. Let's wait for @rchristie 's approval before we merge this in.

File: src/cmlibs/utils/zinc/group.py
Comment (rchristie) [inline src/cmlibs/utils/zinc/group.py:64]: `complement=True`

**Review by rchristie — CHANGES_REQUESTED** [review]
Some improvements and doc recommendations.
Groups are deceptively simple and I'm keen not to have side effects, so these utilities are getting necessarily complicated.
I note one other situation where a side-effect is possible: you can still add an empty mesh group to the target group, which creates an empty mesh group in the target. I think this is OK as caller is unlikely to try to do this.

**Comment by tsalemink** [comment]
@rchristie, thank you for your feedback. I have made all of the changes you requested and I have squashed the commits together.

**Review by rchristie — APPROVED** [review]
Great!

**Pull request merged**
